File: app/src/main/java/openfoodfacts/github/scrachx/openfood/views/adapters/autocomplete/PeriodAfterOpeningAutoCompleteAdapter.java
package openfoodfacts.github.scrachx.openfood.views.adapters.autocomplete;

import android.content.Context;
import android.widget.ArrayAdapter;
import android.widget.Filter;
import android.widget.Filterable;

import androidx.annotation.NonNull;

import org.apache.commons.lang.StringUtils;

import java.util.ArrayList;

import openfoodfacts.github.scrachx.openfood.network.CommonApiManager;
import openfoodfacts.github.scrachx.openfood.network.services.ProductsAPI;

public class PeriodAfterOpeningAutoCompleteAdapter extends ArrayAdapter<String> implements Filterable {
    private ProductsAPI client;
    private final ArrayList<String> periodsList;

    public PeriodAfterOpeningAutoCompleteAdapter(Context context, int textViewResourceId) {
        super(context, textViewResourceId);
        client = CommonApiManager.getInstance().getProductsApi();
        periodsList = new ArrayList<>();
    }

    @Override
    public int getCount() {
        return periodsList.size();
    }

    @Override
    public String getItem(int position) {
        if (position < 0 || position >= periodsList.size()) {
            return StringUtils.EMPTY;
        }
        return periodsList.get(position);
    }

    @NonNull
    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                FilterResults filterResults = new FilterResults();
                filterResults.count = 0;
                if (constraint == null) {
                    return filterResults;
                }
                // Retrieve the autocomplete results from server.
                periodsList.clear();
                periodsList.addAll(client.getPeriodAfterOpeningSuggestions(constraint.toString()).blockingGet());

                // Assign the data to the FilterResults
                filterResults.values = periodsList;
                filterResults.count = periodsList.size();
                return filterResults;
            }

            @Override
            protected void publishResults(CharSequence constraint, FilterResults results) {
                if (results != null && results.count > 0) {
                    notifyDataSetChanged();
                } else {
                    notifyDataSetInvalidated();
                }
            }
        };
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/adapters/autocomplete/PeriodAfterOpeningAutoCompleteAdapter.kt
package openfoodfacts.github.scrachx.openfood.features.adapters.autocomplete

import android.content.Context
import android.widget.ArrayAdapter
import android.widget.Filter
import android.widget.Filterable
import kotlinx.coroutines.runBlocking
import openfoodfacts.github.scrachx.openfood.repositories.ProductRepository
import org.apache.commons.lang3.StringUtils
import java.util.*

class PeriodAfterOpeningAutoCompleteAdapter(
        context: Context,
        textViewResourceId: Int,
        private val client: ProductRepository
) : ArrayAdapter<String>(context, textViewResourceId), Filterable {
    private val periodsList = mutableListOf<String>()

    override fun getCount() = periodsList.size

    override fun getItem(position: Int) =
            if (position < 0 || position >= periodsList.size) StringUtils.EMPTY else periodsList[position]

    override fun getFilter() = object : Filter() {
        override fun performFiltering(constraint: CharSequence?): FilterResults {
            // If no search, 0 results
            if (constraint == null) return FilterResults().apply { count = 0 }

            // Retrieve the autocomplete results from server.
            val list = runBlocking { client.getPeriodAfterOpeningSuggestions(constraint.toString()) }

            // Assign the data to the FilterResults
            return FilterResults().apply {
                values = list
                count = list.size
            }
        }

        override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
            if (results != null && results.count > 0) {
                periodsList.clear()
                periodsList += results.values as ArrayList<String>
                notifyDataSetChanged()
            } else {
                notifyDataSetInvalidated()
            }
        }
    }

}
